Cluster grenade refactor and contra markings

Plykiya opened the pull request:
<!-- Please read these guidelines before opening your PR: https://docs.spacestation14.io/en/getting-started/pr-guideline -->
<!-- The text between the arrows are comments - they will not be visible on your PR. -->

## About the PR
<!-- What did you change in this PR? -->
Cluster grenades essentially have two categories.
- Ones that shoot projectiles (projectile grenades)
- Ones that scatter entities that may potentially trigger themselves (scattering grenades)

I split it into two distinct systems to better separate the logic and fixed a few bugs.

## Why / Balance
<!-- Why was it changed? Link any discussions or issues here. Please discuss how this would affect game balance. -->
- Fixes https://github.com/space-wizards/space-station-14/issues/27571
- Fixes cluster grenades and clusterbanana type grenades not triggering when exploded
- Fixes stingers and incendiaries type grenades not triggering when exploded
- Fixes stacking stingers into stingers into stingers into stingers into stingers

## Technical details
<!-- If this is a code change, summarize at high level how your new code works. This makes it easier to review. -->
Clustergrenades only trigger as a result of their on-use timer. This means that when they were destroyed, nothing happened. I gave both new categories of grenades a damage trigger so that they can be properly chained like regular explosives.

More than half of the variables in the system were unused by projectile-type grenades, so separating the two into two different systems enables easier alteration of logic or behavior.

None of the actual behavior regarding how they function has changed, it's just been separated.

- Scattering grenades and projectile grenades now have an abstract parent to make contra marking sane
- Clusterbang, cluster banana peel, slipocalypse clustersoap, and foam dart grenade migrated to Scattering Grenades and proper contra marking
- Stingers, incendiaries, and shrapnel grenades migrated to Projectile Grenades and proper contra marking

## Media
<!-- 
PRs which make ingame changes (adding clothing, items, new features, etc) are required to have media attached that showcase the changes.
Small fixes/refactors are exempt.
Any media may be used in SS14 progress reports, with clear credit given.

If you're unsure whether your PR will require media, ask a maintainer.
-->

https://private-user-images.githubusercontent.com/58439124/335451517-58b0a78b-48cb-4861-9459-8317323cdf7e.mp4?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjM4NDg3NzcsIm5iZiI6MTcyMzg0ODQ3NywicGF0aCI6Ii81ODQzOTEyNC8zMzU0NTE1MTctNThiMGE3OGItNDhjYi00ODYxLTk0NTktODMxNzMyM2NkZjdlLm1wND9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA4MTYlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwODE2VDIyNDc1N1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWMzMGUyZDY3NmI0ZWY2NzE1NjJhYTRmMzQ2ZDU4ZjRlMGM0YTFiNmMzNDMzMmMxMTcyNDhhZDI4OWU2YzQxOWImWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.Wxrn3HrQTz_lrkifmx16C2U0EvXrKxcKbWic1-V2Juo

https://private-user-images.githubusercontent.com/58439124/335451529-2f9f4d29-1cd1-4fcc-a970-7ae466054ab9.mp4?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjM4NDg3NzcsIm5iZiI6MTcyMzg0ODQ3NywicGF0aCI6Ii81ODQzOTEyNC8zMzU0NTE1MjktMmY5ZjRkMjktMWNkMS00ZmNjLWE5NzAtN2FlNDY2MDU0YWI5Lm1wND9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA4MTYlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwODE2VDIyNDc1N1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTA4NzE4MWQ0NDBjNTJmYWMyZWY2MTg3YjhjMTM5ODQ0MDZiNmNmMDA5MmQxY2I0YjNiMzI0NDhiODZhNzVjMGImWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.sn0_bozS2UE79JWXZcrguA4Uulprqw7u9wWgdsncsZg

https://private-user-images.githubusercontent.com/58439124/341185234-783f1f67-89f2-43bb-9623-166a194ded20.mp4?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjM4NDg3NzgsIm5iZiI6MTcyMzg0ODQ3OCwicGF0aCI6Ii81ODQzOTEyNC8zNDExODUyMzQtNzgzZjFmNjctODlmMi00M2JiLTk2MjMtMTY2YTE5NGRlZDIwLm1wND9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA4MTYlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwODE2VDIyNDc1OFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTcyMDMzMmI1ZTVhZmM5YjhjZWEwOGYyMWRiOWE2OGYxNjQzYjgwZTEyZmVjZjg1MWY5ZDY0ZDVhNzRmN2RiNmMmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.emCSM4pKjC04OB4GUdTZqN3VBLnMgZ-UPn-cI680J-A

## Requirements
<!-- 
Due to influx of PR's we require to ensure that PR's are following the correct guidelines.

Please take a moment to read these if its your first time.

Check the boxes below to confirm that you have in fact seen these (put an X in the brackets, like [X]):
-->
- [x] I have read and I am following the [Pull Request Guidelines](https://docs.spacestation14.com/en/general-development/codebase-info/pull-request-guidelines.html). I understand that not doing so may get my pr closed at maintainer’s discretion
- [x] I have added screenshots/videos to this PR showcasing its changes ingame, **or** this PR does not require an ingame showcase

## Breaking changes
<!--
List any breaking changes, including namespace, public class/method/field changes, prototype renames; and provide instructions for fixing them. This will be pasted in #codebase-changes.
-->
ClusterGrenadeComponent/System were removed, replaced by ProjectileGrenadeComponent/System and ScatteringGrenadeComponent/System.

**Changelog**
<!--
Make players aware of new features and changes that could affect how they play the game by adding a Changelog entry. Please read the Changelog guidelines located at: https://docs.spacestation14.io/en/getting-started/pr-guideline#changelog
-->

<!--
Make sure to take this Changelog template out of the comment block in order for it to show up. Changelog must have a :cl: symbol, so the bot recognizes the changes and adds them to the game's changelog.
:cl:
- add: Added fun!
- remove: Removed fun!
- tweak: Changed fun!
- fix: Fixed fun!
-->
:cl:
 - fix: Bombs like stingers and clustergrenades now trigger when destroyed due to damage.

Plykiya commented:
this is currently possible on live:

https://cdn.discordapp.com/attachments/1191299997489188916/1266110532729966703/Base_Profile_2024.07.25_-_12.10.49.119.mp4?ex=66c0f5d2&is=66bfa452&hm=7a27f8ae6053b5d3de74e74d110b48a7365c8f7a936c5f78586dce5d8b00fccd&

Everturning commented:
I think the stinger hit the nukies boss

Aidenkrz reviewed (approved):
For the love of god the original PR for this was almost 8 months ago, this bug has existed for a millennia at this post and nobody has reviewed this.

Plykiya commented:
900 stingers in a single stinger grenade is probably capable of crashing a server

https://cdn.discordapp.com/attachments/1295923728793866332/1301611809282719824/stinger_grenade.mp4?ex=67509d76&is=674f4bf6&hm=2a6c24eab300c7d9c6287c667a6924fb059d199c2cceb569efdbe4b6b38fc5eb&

chromiumboy commented on Content.Server/Explosion/EntitySystems/ProjectileGrenadeSystem.cs:
```diff
@@ -0,0 +1,111 @@
+            // velocity is randomized to make the projectiles look
+            // slightly even, doesn't really change much but it looks better
+            Vector2 direction = angle.ToVec().Normalized();
+            Vector2 velocity = _random.NextVector2(component.Velocity);
```
So this can generate a value between 0 and component.Velocity? This should have a minimum value

Plykiya replied:
`public void ShootProjectile(EntityUid uid, Vector2 direction, Vector2 gunVelocity, EntityUid gunUid, EntityUid? user = null, float speed = 20f)
    {
        var physics = EnsureComp<PhysicsComponent>(uid);
        Physics.SetBodyStatus(uid, physics, BodyStatus.InAir);

        var targetMapVelocity = gunVelocity + direction.Normalized() * speed;`

even if the provided velocity ends up being 0, apparently the ShootProjectile function has a base speed of 20f plus whatever additional velocity is provided

this does essentially make the component Velocity completely useless... I guess, but it doesn't cause any problems.

chromiumboy commented on Content.Server/Explosion/EntitySystems/ScatteringGrenadeSystem.cs:
```diff
@@ -0,0 +1,178 @@
+    /// There are some scattergrenades you can fill up with more grenades (like clusterbangs)
+    /// This covers how you insert more into it
+    /// </summary>
+    private void OnScatteringInteractUsing(Entity<ScatteringGrenadeComponent> entity, ref InteractUsingEvent args)
```
If possible, this should be in shared so the insertion can be predicted

chromiumboy commented on Content.Server/Explosion/EntitySystems/ScatteringGrenadeSystem.cs:
```diff
@@ -0,0 +1,178 @@
+            var totalCount = component.Container.ContainedEntities.Count + component.UnspawnedCount;
+
+            // if triggered while empty, (if it's blown up while empty) it'll just delete itself
+            if (component.IsTriggered)
+            {
+                if (totalCount > 0)
```
```suggestion
            if (component.IsTriggered && totalCount > 0)
```

chromiumboy replied:
The ending brace will need to be deleted as well

chromiumboy reviewed (changes requested):
Thanks for this! Just a few things from me

Also, during my tests I noted that when projectile grenades, like stingers, are triggered it causes a whole bunch of errors because the meta data component on the grenade can't be resolved. This is likely because it got deleted too early. So that needs to be addressed as well

Plykiya commented:
> Thanks for this! Just a few things from me
> 
> Also, during my tests I noted that when projectile grenades, like stingers, are triggered it causes a whole bunch of errors because the meta data component on the grenade can't be resolved. This is likely because it got deleted too early. So that needs to be addressed as well

because the bullet considers the grenade shooting it the "gun" and the grenade gets deleted so there is technically no gun, yeah

chromiumboy commented:
> > Thanks for this! Just a few things from me
> > Also, during my tests I noted that when projectile grenades, like stingers, are triggered it causes a whole bunch of errors because the meta data component on the grenade can't be resolved. This is likely because it got deleted too early. So that needs to be addressed as well
> 
> because the bullet considers the grenade shooting it the "gun" and the grenade gets deleted so there is technically no gun, yeah

Are you able to fix it? All those errors cause noticeable lag

slarticodefast commented:
I'm pretty sure this is the same as this bug already happening on master
#32642

chromiumboy commented:
> I'm pretty sure this is the same as this bug already happening on master #32642

Yeah, it's the same issue

Plykiya commented:
https://github.com/user-attachments/assets/bd42933c-84c8-47ca-8273-ce5698888ff0

Plykiya commented:
looks pretty good

Plykiya commented:
https://github.com/user-attachments/assets/5e0d6d74-e48b-42a3-b776-a458ae9f75c1

chromiumboy commented on Content.Shared/Explosion/EntitySystems/SharedScatteringGrenadeSystem.cs:
```diff
@@ -0,0 +1,70 @@
+
+namespace Content.Shared.Explosion.EntitySystems;
+
+public sealed class SharedScatteringGrenadeSystem : EntitySystem
```
```suggestion
public abstract class SharedScatteringGrenadeSystem : EntitySystem
```

chromiumboy commented on Content.Server/Explosion/EntitySystems/ScatteringGrenadeSystem.cs:
```diff
@@ -0,0 +1,125 @@
+
+namespace Content.Server.Explosion.EntitySystems;
+
+public sealed class ScatteringGrenadeSystem : EntitySystem
```
```suggestion
public sealed class ScatteringGrenadeSystem : SharedScatteringGrenadeSystem 
```

chromiumboy commented on Content.Server/Explosion/EntitySystems/ScatteringGrenadeSystem.cs:
```diff
@@ -0,0 +1,125 @@
+﻿using Content.Shared.Explosion.Components;
+using Content.Shared.Interaction;
+using Content.Shared.Throwing;
+using Content.Shared.Whitelist;
```
```suggestion
```

chromiumboy commented on Content.Server/Explosion/EntitySystems/ScatteringGrenadeSystem.cs:
```diff
@@ -0,0 +1,125 @@
+﻿using Content.Shared.Explosion.Components;
+using Content.Shared.Interaction;
```
```suggestion
```

chromiumboy commented on Content.Server/Explosion/EntitySystems/ScatteringGrenadeSystem.cs:
```diff
@@ -0,0 +1,125 @@
+    [Dependency] private readonly IRobustRandom _random = default!;
+    [Dependency] private readonly ThrowingSystem _throwingSystem = default!;
+    [Dependency] private readonly TransformSystem _transformSystem = default!;
+    [Dependency] private readonly SharedScatteringGrenadeSystem _scatteringGrenadeSystem = default!;
```
```suggestion
```

chromiumboy commented on Content.Server/Explosion/EntitySystems/ProjectileGrenadeSystem.cs:
```diff
@@ -0,0 +1,111 @@
+using Robust.Shared.Containers;
+using Robust.Shared.Map;
+using Robust.Shared.Random;
+using System.Numerics;
```
```suggestion
```

chromiumboy commented on Content.Shared/Explosion/Components/ScatteringGrenadeComponent.cs:
```diff
@@ -0,0 +1,108 @@
+/// Use this component if the grenade splits into entities that make use of Timers
+/// or if you just want it to throw entities out in the world
+/// </summary>
+[RegisterComponent, NetworkedComponent, AutoGenerateComponentState]
```
```suggestion
[RegisterComponent, NetworkedComponent, AutoGenerateComponentState, Access(typeof(SharedScatteringGrenadeSystem))]
```

chromiumboy commented on Content.Shared/Explosion/Components/ScatteringGrenadeComponent.cs:
```diff
@@ -0,0 +1,108 @@
+﻿using Content.Shared.Whitelist;
```
```suggestion
using Content.Shared.Explosion.EntitySystems;
﻿using Content.Shared.Whitelist;
```

chromiumboy reviewed (changes requested):
Just a few more things

chromiumboy commented:
It doesn't look like the changes to RequireProjectileTargetSystem.cs have stopped the errors when projectile grenades detonate, unfortunately. I had a go at fixing it, but I couldn't make any progress either

Plykiya commented:
> It doesn't look like the changes to RequireProjectileTargetSystem.cs have stopped the errors when projectile grenades detonate, unfortunately. I had a go at fixing it, but I couldn't make any progress either

it stopped the client-side errors that normally caused me lag at the very least

chromiumboy reviewed (approved):
Looks good :)

Plykiya commented:
to anyone who's wondering what the issue is for the future

SharedGunSystem's ShootProjectile() sets the shooter to either the gun's EntityUid, or a nullable user EntityUid using Projectiles.SetShooter(). This will get set to our projectile grenade's EntityUid. The projectile grenade is deleted once triggered.

ProjectileComponent has an EntityUid? Shooter field that is AutoNetworked. This field is set to the now deleted projectile grenade's EntityUid, and that will cause the engine's attempt to automatically network the EntityUid to fail since it can't load the MetadataComponent from the now deleted EntityUid

Avalon-Proto commented:
I got a question, I've been attempting to make a grenade launcher that fires timed grenades in another server. Think this will help me out?

Plykiya commented:
> I got a question, I've been attempting to make a grenade launcher that fires timed grenades in another server. Think this will help me out?

that's exactly why I made this

Avalon-Proto commented:
> > I got a question, I've been attempting to make a grenade launcher that fires timed grenades in another server. Think this will help me out?
> 
> that's exactly why I made this change, so I could make the china-lake fire different types of ammo like that lol

THANK YOU. I can now give Delta's secborgs teargas and stingers instead of a pack injector

Avalon-Proto commented:
So, I have gotten this PR of mine all working, except for 1 thing, the grenades launched by my launcher do not start the timer. How canI go abotu fixing that?
https://github.com/DeltaV-Station/Delta-v/pull/2059
